stash/branch.go

```diff
@@ -0,0 +1,175 @@
+	return nil
+}
+
+func (r *Branch) Equals(actual gitprovider.InfoRequest) bool {
+	return reflect.DeepEqual(r, actual)
+}
+
+func NewStashBranches(client *stashClientImpl) StashBranches {
+	p := &Branches{
+		Paging:    Paging{},
+		Requester: client.Client(),
+		Branches:  make([]*Branch, 0),
+		log:       client.log,
+	}
+	return p
+}
+
+type StashBranches interface {
+	List(ctx context.Context, projectName, repoName string, opts *ListOptions) (*Paging, error)
+	Get(ctx context.Context, projectName, repoName, branchID string) (*Branch, error)
+	Default(ctx context.Context, projectName, repoName string) (*Branch, error)
+	getBranches() []*Branch
+}
+
+type Branches struct {
+	StashBranches
+	Paging
+	httpclient.Requester
+	Branches []*Branch `json:"values,omitempty"`
+	log      logr.Logger
+}
+
+func (p *Branches) getBranches() []*Branch {
+	return p.Branches
+}
+
+func (p *Branches) List(ctx context.Context, projectName, repoName string, opts *ListOptions) (*Paging, error) {
+	var query *url.Values
+	query = addPaging(query, opts)
+	resp, err := p.Requester.Do(ctx, newURI(projectsURI, projectName, RepositoriesURI, repoName, branchesURI), query, nil, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("list branches for repository failed, %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("list branches for repository api call returned unexpected status code: %d", resp.StatusCode)
+	}
+
+	respBody, err := httpclient.GetRespBody(resp)
+	if err != nil {
+		return nil, fmt.Errorf("list branches for repository failed, unable to obtain response body, %w", err)
+	}
+
+	if err := json.Unmarshal([]byte(respBody), &p); err != nil {
+		return nil, fmt.Errorf("list branches for repository failed, unable to unmarshall repository json, %w", err)
+	}
+
+	for _, branches := range p.getBranches() {
+		branches.setSessionInfo(resp)
+	}
+	return &p.Paging, nil
+}
+
+func (p *Branches) Get(ctx context.Context, projectName, repoName, branchID string) (*Branch, error) {
+	var query *url.Values
+	query = addPaging(query, &ListOptions{})
+	resp, err := p.Requester.Do(ctx, newURI(projectsURI, projectName, RepositoriesURI, repoName, branchesURI), query, nil, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("get branch failed, %w", err)
+	}
+
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("get branch for repository api call returned unexpected status code: %d", resp.StatusCode)
+	}
+
+	respBody, err := httpclient.GetRespBody(resp)
+	if err != nil {
+		return nil, fmt.Errorf("get branch failed, unable to obtain response body, %w", err)
+	}
+
+	if err := json.Unmarshal([]byte(respBody), &p); err != nil {
+		return nil, fmt.Errorf("get branch for repository failed, unable to unmarshall repository json, %w", err)
+	}
+
+	for _, branch := range p.getBranches() {
+		if branch.DisplayId == branchID {
+			branch.setSessionInfo(resp)
+			return branch, nil
+		}
+	}
+
+	return nil, gitprovider.ErrNotFound
```

Comment:
I think it should only return `ErrNotFound` when the server responds with a 404 response.

Otherwise people spend time debugging the wrong things.